man/stack/sp_stack_print.3
.\"  Staple - A general-purpose data structure library in pure C89.
.\"  Copyright (C) 2021  Randoragon
.\"
.\"  This library is free software; you can redistribute it and/or
.\"  modify it under the terms of the GNU Lesser General Public
.\"  License as published by the Free Software Foundation;
.\"  version 2.1 of the License.
.\"
.\"  This library is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\"  Lesser General Public License for more details.
.\"
.\"  You should have received a copy of the GNU Lesser General Public
.\"  License along with this library; if not, write to the Free Software
.\"  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
.\"--------------------------------------------------------------------------------
.TH SP_STACK_PRINT 3 DATE "libstaple-VERSION"
.SH NAME
sp_stack_print,
sp_stack_printc,
sp_stack_prints,
sp_stack_printi,
sp_stack_printl,
sp_stack_printsc,
sp_stack_printuc,
sp_stack_printus,
sp_stack_printui,
sp_stack_printul,
sp_stack_printf,
sp_stack_printd,
sp_stack_printld,
sp_stack_printll,
sp_stack_printull,
sp_stack_printu8,
sp_stack_printu16,
sp_stack_printu32,
sp_stack_printu64,
sp_stack_printi8,
sp_stack_printi16,
sp_stack_printi32,
sp_stack_printi64,
sp_stack_printstr
\- print contents of a stack
.SH SYNOPSIS
.ad l
#include <staple.h>
.sp
int
.BR sp_stack_print "(const struct sp_stack"
.RI * stack ,
int
.RI (* func ")(const void*))"
.br
int
.BR sp_stack_printc "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_prints "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printi "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printl "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printsc "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printuc "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printus "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printui "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printul "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printf "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printd "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printld "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printll "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printull "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printu8 "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printu16 "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printu32 "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printu64 "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printi8 "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printi16 "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printi32 "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printi64 "(const struct sp_stack"
.RI * stack )
.br
int
.BR sp_stack_printstr "(const struct sp_stack"
.RI * stack )
.sp
Link with \fI-lstaple\fP.
.ad
.SH DESCRIPTION
.P
The functions in the
.BR sp_stack_print ()
family are used to print useful information about a stack, and are
provided for debugging purposes.
.P
.I stack
is the stack that shall be printed.
.SS Generic Form
.I func
is an optional pointer to a custom print function. The function must be
implemented in such a way that it accepts the address of
an element as its argument.
.P
The function pointed to by
.I func
must return 0 for success, and anything else for failure. In debug mode, the
caught error code (if any) is printed on stderr.
.P
.B NULL
can be passed as
.IR func ,
in which case
.BR sp_stack_print ()
will default to printing the address of each element.
.SH RETURN VALUE
If successful, the functions in the
.BR sp_stack_print ()
family shall return 0. Otherwise, they shall return a code indicating the
error.
.SH ERRORS
The functions in the
.BR sp_stack_print ()
family shall fail if:
.IP \fBSP_EINVAL\fP 1.5i
.I stack
is a null pointer (debug mode only).
.SS Generic Form
The
.BR sp_stack_print ()
function shall fail if:
.IP \fBSP_ECALLBK\fP 1.5i
.I func
returned non-0 for some stack element.
.SS Suffixed Form
The suffixed
.BR sp_stack_print ()
functions shall fail if:
.IP \fBSP_EILLEGAL\fP 1.5i
.IR stack->elem_size
does not match the size of the function's return type (debug mode only).
.SH CONFORMING TO
See
.BR libstaple (7).
.SH SEE ALSO
.ad l
.BR libstaple (7),
.BR sp_stack (7),
.BR sp_stack_create (3),
.BR sp_stack_destroy (3),
.BR sp_stack_clear (3),
.BR sp_stack_push (3),
.BR sp_stack_peek (3),
.BR sp_stack_pop (3),
.BR sp_stack_insert (3),
.BR sp_stack_remove (3),
.BR sp_stack_qinsert (3),
.BR sp_stack_qremove (3),
.BR sp_stack_get (3),
.BR sp_stack_set (3),
.BR sp_stack_eq (3),
.BR sp_stack_copy (3),
.BR sp_stack_map (3)


src/stack/sp_stack_print.c
/*  Staple - A general-purpose data structure library in pure C89.
 *  Copyright (C) 2021  Randoragon
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation;
 *  version 2.1 of the License.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "../sp_stack.h"
#include "../internal.h"
#include "../sp_errcodes.h"

int sp_stack_print(const struct sp_stack *stack, int (*func)(const void*))
{
	size_t i;
#ifdef STAPLE_DEBUG
	if (stack == NULL) {
		error(("stack is NULL"));
		return SP_EINVAL;
	}
#endif
	printf("sp_stack_print()\nsize/capacity: "SP_SIZE_FMT"/"SP_SIZE_FMT", elem_size: "SP_SIZE_FMT"\n",
		(SP_SIZE_T)stack->size, (SP_SIZE_T)stack->capacity, (SP_SIZE_T)stack->elem_size);
	if (func == NULL)
		for (i = stack->size; i-- > 0;) {
			const void *const elem = (char*)stack->data + i * stack->elem_size;
			printf("["SP_SIZE_FMT"]\t%p\n", (SP_SIZE_T)stack->size - 1 - i, elem);
		}
	else
		for (i = stack->size; i-- > 0;) {
			const void *const elem = (char*)stack->data + i * stack->elem_size;
			int err;
			printf("["SP_SIZE_FMT"]\t", (SP_SIZE_T)stack->size - 1 - i);
			if ((err = func(elem)) != 0) {
				error(("callback function func returned %d (non-0)", err));
				return SP_ECALLBK;
			}
		}
	return 0;
}
